### ISSUE TITLE
Consider refactor with keras style

### ISSUE THREAD

**Issue by pnrobinson** [monarch-initiative/N2V#14]
See Part II: Writing your own training & evaluation loops from scratch
https://www.tensorflow.org/guide/keras/train_and_evaluate

**Comment by callahantiff** [comment]
I see that the `NCE` class in `kW2V.py` is based off of [this](https://www.eggie5.com/134-nce-Noise-contrastive-Estimation-Loss#Keras).

This is great, but I'm also wondering if we want to try and actually follow the keras typing more closely? This is motivated by having read the Hands on ML book, maybe a bit too closely :P.

Anywhoo, other thoughts that come to mind for things to think about and also things we can talk about today:
- Utilizing the [`keras.preprocessing.sequence.skipgrams()`](https://keras.io/preprocessing/sequence/#skipgrams)  
  - Would allows us to take advantage of the word rank-based probabilistic sampling table this method calls, or at least leverage parts of it and extend the bits we don't like.  

Will keep adding as I go...

**Comment by callahantiff** [comment]
Useful example of keras implementation of skipgram and cbow [here](https://towardsdatascience.com/understanding-feature-engineering-part-4-deep-learning-methods-for-text-data-96c44370bbfa).

**Comment by callahantiff** [comment]
@pnrobinson - here is the rough outline of what I was thinking for the more general embedding implementation. Looking forward to discussing this afternoon!

<br>

____

**Inspiration:** [Towards Data Science Article](https://towardsdatascience.com/understanding-feature-engineering-part-4-deep-learning-methods-for-text-data-96c44370bbfa)

### General Design
Separate out running the model (i.e. actual call to implement the model training) from the tasks that are needed to run a model (i.e. customized training, skip-gram generator, batch sampler). A third script, would be created to store different kinds of classes that subclass existing `keras` classes. The hope is that by separating out these components it will be easy to add new features (i.e. models, loss functions, metrics) as they are needed without having to significantly overhaul existing code. I also think that this design would allow us to take advantage of the code that has already been written.

Currently, there would be three classes, which are needed in order to run an embedding model. Each class is listed below with a general description. Names can change, just using these examples.

**Scripts:**  
- `ModelHelper()`   
- `KerasExtender()`  
- `EmbeddingModeler()`

<br>

#### `ModelHelper()`
Stores all custom functions needed to run an embedding model.

Examples:
- custom cbow batch function/context pair generator (from  [Towards Data Science Article](https://towardsdatascience.com/understanding-feature-engineering-part-4-deep-learning-methods-for-text-data-96c44370bbfa))
- Tokenizer method
- Skip-gram generator (from  [Towards Data Science Article](https://towardsdatascience.com/understanding-feature-engineering-part-4-deep-learning-methods-for-text-data-96c44370bbfa))
- Model visualizer 
- Load pretrained models
- Load existing embeddings

<br>

#### `KerasExtender()`
Stores all subclassed `keras` classes needed for running a variety of embedding models. As noted in keras documentation, the script could be divided into separate classes that store subclassed versions of existing losses, metrics, initializers, and constraints classes. 

Examples:
 - `Losses()` - keras-subclassed `NCE()` loss function

<br>

#### `EmbeddingModeler()`

General Work Flow:
- Run an embedding model specified by user
- Call `EmbeddingHelper()` to access custom methods for tokenization/training
- Call `KerasExtensions()` to access subclassed keras classes
- Train model
- Store and export training performance (loss)
- Save model and embeddings


*EMBEDDING MODELS*

**CBOW**
*inputs:* `vocab_size`, `embed_size`, `window_size`, `activation`, `loss`, `optimizier`

*model core:*
```python
cbow = Sequential()
cbow.add(Embedding(input_dim=`vocab_size`, output_dim=`embed_size`, input_length=`window_size*2`))
cbow.add(Lambda(lambda x: K.mean(x, axis=1), output_shape=(`embed_size`,)))
cbow.add(Dense(`vocab_size`, activation=`activation`))
cbow.compile(loss=`loss`, optimizer=`optimizier`)

```

<br>

**Skip-gram**
*inputs:* `vocab_size`, `embed_size`, `embeddings_initializer`, `input_length`, `kernel_initializer`, `activation`, `loss`, `optimizier`

*model core:*

```python
# model for target word
word_model = Sequential()
word_model.add(Embedding(`vocab_size`, `embed_size,`
                         embeddings_initializer=`embeddings_initializer`,
                         input_length=`1`))
word_model.add(Reshape((`embed_size`, )))

# model for context word
context_model = Sequential()
context_model.add(Embedding(`vocab_size`, `embed_size,`
                  embeddings_initializer=`embeddings_initializer`,
                  input_length=`1`))
context_model.add(Reshape((`embed_size`,)))

# merge models
model = Sequential()
model.add(Merge([word_model, context_model], mode="dot"))
model.add(Dense(1, kernel_initializer=`kernel_initializer`, activation=`activation`))
model.compile(loss=`loss`, optimizer=`optimizier`")

```

<br>

_Other Models_

**Gensim word2vec**

**GloVE**

**FastText**

**sent2vec**

***


Other scripts could then be written for methods to manipulate, analyze and visualize embeddings (e.g. TensorBoard, t-SNE, etc).

**Comment by callahantiff** [comment]
Future iterations might want to consider replacing current functionality for importing and manipulating data with [`tf.data` API](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/data). Here are some good examples of recent uses:  
  - Comparison of [`Numpy` vs. `tf.data` API](https://medium.com/swlh/data-pipeline-using-tensorflow-dataset-api-with-keras-fit-generator-8c7a3e01c4fd)   
    - Note that this tutorial rovides examples in tf 1 and tf 2

**Comment by callahantiff** [comment]
and saving models: https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/saved_model

**Comment by justaddcoffee** [comment]
Here's a project for this Keras refactor, per discussion with Tiffany/Peter:
https://github.com/monarch-initiative/N2V/projects/1

**Comment by justaddcoffee** [comment]
Tiffany's suggestion: 3 steps for refactor
 1) data preprocessing  using tensor flow/keras framework
2) batching  using tf/keras framework
3) subclass existing classes to add new things

**Comment by justaddcoffee** [comment]
@pnrobinson @vidarmehr Deepak I think is going to make a gitter channel specifically for this refactor - I think he'll add you both to this channel

**Comment by callahantiff** [comment]
> Tiffany's suggestion: 3 steps for refactor
>  1) data preprocessing  using tensor flow/keras framework
> 2) batching  using tf/keras framework
> 3) subclass existing classes to add new things

Thanks for adding this @justaddcoffee, we could consider adding these steps too:

4) fit/training embedding models
5) do stuff with embedding model output (i.e. link prediction, classification, etc)

I'm happy to focus on tf.kerasizing the preprocessing stuff, do you want me to create a new card in the project board for this or would you rather I create a separate issue for that task and include it in the board that way?

**Comment by justaddcoffee** [comment]
I think add a card for this ( 1) data preprocessing  using tensor flow/keras framework) in the project board

**Comment by justaddcoffee** [comment]
I'll add my bit to the project board too (e2e test that encapsulates Vida/Peter's python notebooks)

**Comment by justaddcoffee** [comment]
> I think add a card for this ( 1) data preprocessing  using tensor flow/keras framework) in the project board

@callahantiff actually just added this card for you (and one for me too)

**Comment by callahantiff** [comment]
Happy Saturday!

Having had some more time to read `tf` documentation as well as having had the time to go through the `csf_graph.py` script, I think there is a decision to be made about when in the current workflow we want to convert our data to a `tf.Dataset` object. From my reading/Googling, I think we have two options (with a bit more information provided for Option 2):

**OPTION 1:** we pursue a more extensive input pipeline that uses the `tf.Data` functions from the time that the data is read in. I have explored some of this and it seems reasonable. It will primarily rely on applying iterators to manipulate and update the data. 

For example, we would begin by reading in the data to a [`tf.data.TextLineDataset`](https://www.tensorflow.org/api_docs/python/tf/data/TextLineDataset?hl=en):

```python
filepath = 'tests/data/small_graph.txt'

for line in tf.data.TextLineDataset([filepath]):
    fields = line.split()
    node_a = fields[0]
    node_b = fields[1]
...

```

All of the objects that are updated as the data is read in would be then be `tf.tf.data.Dataset` or `tf.Tensor`.

<br>

**OPTION 2:** we convert our `data` object to a `tf.Dataset` immediately following the bulk of preprocessing (i.e. after performing tokenization if data is text or after creating the CSF graph if the data is a knowledge graph), so that when those scripts finish running they output `tf.Dataset` objects. If we go this route, it would primarily mean updating the `data` or `walks` object that is output by the `csf_graph.py`, `text_encoder.py`, and `hetnode2vec.py` scripts. For example:

`Sentences/Walks (list or nested list):` A list of words, list of word lists of words, or list of node lists, for example: 

```python
walks = [ [8, 4, 3, 7, 3, 2, 1, 0, 1, 6],
          [7, 3, 5, 9, 10, 6, 10, 8, 7, 4],
          [0, 5, 0, 2, 1, 2, 0, 1, 0, 1],
          [10, 8, 9, 7, 9, 7, 3, 2, 3, 7]
         ]
```

_Conversion to TensorFlow Objects_:  
Convert list or nested list to [`tf.data.Dataset.from_tensor_slices`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset), example shown below:

```python
tf_walks = tf.data.Dataset.from_tensor_slices(walks)

for lst in walks:
    print(lst)

tf.Tensor([8 4 3 7 3 2 1 0 1 6], shape=(10,), dtype=int32)
tf.Tensor([ 7  3  5  9 10  6 10  8  7  4], shape=(10,), dtype=int32)
tf.Tensor([0 5 0 2 1 2 0 1 0 1], shape=(10,), dtype=int32)
tf.Tensor([10  8  9  7  9  7  3  2  3  7], shape=(10,), dtype=int32)

```

<br>

**Question:** What do you guys think? I am more than happy to make the changes to implement either option, but I think this is a good time to get some feedback from the rest of group 😄.

**Comment by deepakunni3** [comment]
@callahantiff If I understand the options correctly, Option 1 would require a bit of rewrite on how we currently parse the input data. But this approach seems more efficient since we are avoiding the need to create intermediate data objects. 
Are the intermediates being used downstream? If not, then perhaps it would make sense to go with option 1.

**Comment by callahantiff** [comment]
> @callahantiff If I understand the options correctly, Option 1 would require a bit of rewrite on how we currently parse the input data. But this approach seems more efficient since we are avoiding the need to create intermediate data objects.
> Are the intermediates being used downstream? If not, then perhaps it would make sense to go with option 1.

Thanks for the feedback @deepakunni3! I agree with your points about avoiding creating extra intermediate data objects. Currently, there are several preprocessing steps performed to create these objects, prior to when they are eventually used for batching/training. The sentences/walks objects are used downstream when training a cbow and/or skipgram model. 

I also just came across [this](https://medium.com/@prasad.pai/how-to-use-dataset-and-iterators-in-tensorflow-with-code-samples-3bb98b6b74ab) article, which while focused on `tf1`, discusses the correct usage of `tf.Data` objects that exist in `tf2`. It looks like it will be helpful.

OK, I am planning to work on this again on Wednesday. I can start trying to make the changes then, unless there are any other concerns with proceeding in this manner?

**Comment by justaddcoffee** [comment]
I think either option 1 or 2 seems reasonable, and also I think you (Tiffany) are possibly in a good position to make this call. 

For my $0.02, it seems like option 2 would both be easier to implement, and also would give us more flexibility when implementing new algorithms, since they wouldn't be as tightly coupled to Tf. I think either is reasonable though...

**Comment by callahantiff** [comment]
> I think either option 1 or 2 seems reasonable, and also I think you (Tiffany) are possibly in a good position to make this call.
> 
> For my $0.02, it seems like option 2 would both be easier to implement, and also would give us more flexibility when implementing new algorithms, since they wouldn't be as tightly coupled to Tf. I think either is reasonable though...

@justaddcoffee - Great points as well. OK, I think I will take a fresh look Wednesday morning and then just make a decision. I'll update here afterwards.

Thanks so much to you @deepakunni3 for the feedback, I appreciate it very much!

**Comment by deepakunni3** [comment]
@justaddcoffee Yes, thats a good point regarding coupling with TensorFlow.
I would be curious what the consensus is on using TensorFlow for parsing input data, and 
batch using iterators, etc; even for algorithms that are newly implemented.  @pnrobinson @vidarmehr 
If there is a unified framework for building out these datasets, it would be easier to handle and debug since all algorithms will be relying on the same data structure.

@callahantiff Thanks for listing out the possibilities. This is very helpful! Looking forward to see what you find.

**Comment by pnrobinson** [comment]
I would suggest that probably there are good Option 1-options in TF for pure text processing that we should use. For the random walk, Option-2 is probably the only realistic option, and it is unclear that changing the random walk code to use tensors would buy us any additional performance anyway!

**Comment by callahantiff** [comment]
> I would suggest that probably there are good Option 1-options in TF for pure text processing that we should use. For the random walk, Option-2 is probably the only realistic option, and it is unclear that changing the random walk code to use tensors would buy us any additional performance anyway!

@pnrobinson - Wanted to send you a quick update after the call today.

During the meeting we tentatively decided that we would move forward with [option 2](https://github.com/monarch-initiative/N2V/issues/14#issuecomment-586664264) because we think that this will make it easier to add or extend different graph (and text-based) structures and/or preprocessing functions in the future.

**Plan going forward:** Duplicate current scripts on the `creating_keras_develop_branch` (appending `_tf` to the end) and begin working through each, converting the functions to take `tf.Tensor` objects. The thought being, that by duplicating the scripts we would not negatively impact existing workflows that are already working great. 

**Does this plan sound OK to you?**

**Comment by callahantiff** [comment]
Quick Update from this week's call:

I have added a new argument to `word2vec.py` which enables the user to specify whether or not to run computations over GPUs or CPUs called `device_type`, which takes values "cpu" or "gpu" (i.e. `self.device_type = '/CPU:0' if 'cpu' in device_type.lower() else '/GPU:0'`). The default is CPU. We can adjust this further in the future, but currently the algorithm will use the first CPU or GPU visible by `tf`, which seems to be pretty standard from what I can tell .

**Comment by pnrobinson** [comment]
@callahantiff @deepakunni3 @vidarmehr 
After studying TF2 for a few more weeks, I do not think that we will be able to do this by subclassing tf.keras.models.Model, at least not without reimplementing a lot of other things. For instance, the NCE loss function is provided by tf2, but it needs weights and biases of the output layer as inputs. While we can do this, the resulting code is not that easy to understand.
Unless somebody sees a clear way forward, I would propose that we re-write the batching functions using the tf.data.Datasets code from the colaboratory and try to merge all of the branches this week. I think we are pretty close with GloVe, and if we can manage to integrate all of that this week, we could test next week. With Leo, I am trying to get all of the syn lethal datasets, and so we could do the final analysis hopefully within a month?

**Comment by callahantiff** [comment]
@pnrobinson - I think that sounds completely reasonable and makes sense, especially given our timeline.

**Comment by pnrobinson** [comment]
So does this mean mainly starting from creating_keras_develop_branch ?
The main changes in this branch are for the CSFGraph that it returns tensors?
Should we also add a class to input texts?

**Comment by callahantiff** [comment]
Yes, I believe so.

I have also updated the [`text_encoder_tf.py`](https://github.com/monarch-initiative/N2V/blob/creating_keras_develop_branch/xn2v/text_encoder_tf.py) script. It might provide a start for a text class or perhaps we want a fresh start using the workflows from the Coursera course that you took?

**Comment by pnrobinson** [comment]
closed by https://github.com/monarch-initiative/N2V/issues/127
